File: solution/java/day14/src/main/java/games/FizzBuzz.java
package games;

import java.util.LinkedHashMap;
import java.util.Map;

import static games.Result.failure;
import static games.Result.fromSuccess;

public class FizzBuzz {
    private static final int MIN = 1;
    private static final int MAX = 100;
    private static final int FIZZ = 3;
    private static final int BUZZ = 5;
    private static final int FIZZBUZZ = 15;

    private static final Map<Integer, String> mapping;

    static {
        mapping = new LinkedHashMap<>();
        mapping.put(FIZZBUZZ, "FizzBuzz");
        mapping.put(FIZZ, "Fizz");
        mapping.put(BUZZ, "Buzz");
    }

    public static Result<String> convert(int input) {
        return isOutOfRange(input)
                ? failure()
                : fromSuccess(convertSafely(input));
    }

    private static String convertSafely(Integer input) {
        return mapping.entrySet()
                .stream()
                .filter(f -> is(f.getKey(), input))
                .findFirst()
                .map(Map.Entry::getValue)
                .orElse(input.toString());
    }

    private static boolean is(Integer divisor, Integer input) {
        return input % divisor == 0;
    }

    private static boolean isOutOfRange(Integer input) {
        return input < MIN || input > MAX;
    }
}

File: solution/kotlin/day14/src/main/kotlin/games/FizzBuzz.kt
package games

import arrow.core.None
import arrow.core.Option
import arrow.core.Some

private const val MIN = 0
private const val MAX = 100
private const val FIZZBUZZ = 15
private const val FIZZ = 3
private const val BUZZ = 5

object FizzBuzz {
    fun convert(input: Int): Option<String> = when {
        isOutOfRange(input) -> None
        else -> Some(convertSafely(input))
    }

    private fun convertSafely(input: Int): String = when {
        `is`(FIZZBUZZ, input) -> "FizzBuzz"
        `is`(FIZZ, input) -> "Fizz"
        `is`(BUZZ, input) -> "Buzz"
        else -> input.toString()
    }

    private fun `is`(divisor: Int, input: Int): Boolean = input % divisor == 0
    private fun isOutOfRange(input: Int) = input <= MIN || input > MAX
}